app/shared/cms/haldoc.ts

```diff
@@ -1,98 +1,165 @@
-import {Http, Headers} from 'angular2/http';
 import {Observable} from 'rxjs/Observable';
-import TemplateParser from 'url-template';
+import {HalRemote} from './halremote';
+
+// Bring observables up to snuff
+export interface HalObservable<T> extends Observable<T> {
+  follow(rel: string, params?: {}): HalObservable<T>;
+  followList(rel: string, params?: {}): HalObservable<T[]>;
+  followItems(rel: string, params?: {}): HalObservable<T[]>;
+}
+Observable.prototype['follow'] = function(rel: string, params: {} = null) {
+  return this.flatMap((doc: HalDoc) => {
+    return doc.follow(rel, params);
+  });
+};
+Observable.prototype['followList'] = function(rel: string, params: {} = null) {
+  return this.flatMap((doc: HalDoc) => {
+    return doc.followList(rel, params);
+  });
+};
+Observable.prototype['followItems'] = function(rel: string, params: {} = null) {
+  return this.flatMap((doc: HalDoc) => {
+    return doc.followItems(rel, params);
+  });
+};
 
 /*
  * Generic class for interacting with HAL api
  */
 export class HalDoc {
-  private http: Http;
-  private host: string;
-  private token: string;
-
-  constructor(data: any = {}, http: Http, host: string, authToken: string) {
-    Object.keys(data).forEach((key) => { this[key] = data[key]; });
-    this.http = http;
-    this.host = host;
-    this.token = authToken;
+
+  constructor(data: any = {}, protected remote: HalRemote) {
+    this.setData(data);
   }
 
-  followLink(linkObj: any, params: any = {}): Observable<HalDoc> {
-    let headers = new Headers();
-    headers.append('Accept', 'application/json');
-    if (this.token) {
-      headers.append('Authorization', `Bearer ${this.token}`);
+  update(data: any): HalObservable<HalDoc> {
+    let link = this['_links'] ? this['_links']['self'] : null;
+    if (!link) {
+      return <HalObservable<HalDoc>> this.error(`Expected update link at _links.self - got null`);
```

Comment:
Here's a consequence of using `HalObservable` - things like `Observable.empty()` and `Observable.error()` also extend Observable.  So you can't directly return an `Observable.error()` without typescript complaining that it can't be cast to a `HalObservable`.

My hack was to return the error from a different method and then cast it.  Makes typescript happy.